server/coordinator/procedure/shard/shard_fsm.go

```diff
@@ -0,0 +1,229 @@
+	}
+
+	if err := updateShardRole(ctx, c, oldLeaderShardID, clusterpb.ShardRole_FOLLOWER); err != nil {
+		event.Cancel(errors.Wrap(err, EventTransferFollower))
+	}
+}
+
+func transferFollowerFailedCallback(event *fsm.Event) {
+	request := event.Args[0].(*LeaderCallbackRequest)
+	ctx := request.Ctx
+	c := request.C
+	actionDispatch := request.ActionDispatch
+	oldLeaderNode := request.OldLeaderNode
+	oldLeaderShardID := request.OldLeaderShardID
+
+	// Transfer failed, stop transfer and reset state
+	action := dispatch.OpenShardAction{ShardIDs: []uint32{oldLeaderShardID}}
+	if err := actionDispatch.OpenShards(ctx, oldLeaderNode, action); err != nil {
+		event.Cancel(errors.Wrap(err, EventTransferFollowerFailed))
+	}
+
+	if err := updateShardRole(ctx, c, oldLeaderShardID, clusterpb.ShardRole_LEADER); err != nil {
+		event.Cancel(errors.Wrap(err, EventTransferFollowerFailed))
+	}
+}
+
+func prepareTransferLeaderCallback(event *fsm.Event) {
+	request := event.Args[0].(*FollowerCallbackRequest)
+	ctx := request.Ctx
+	c := request.C
+	newLeaderShardID := request.NewLeaderShardID
+
+	// TODO: add ShardRole_PENDING_LEADER enum, replace ShardRole_LEADER to ShardRole_PENDING_LEADER
+	if err := updateShardRole(ctx, c, newLeaderShardID, clusterpb.ShardRole_LEADER); err != nil {
+		event.Cancel(errors.Wrap(err, EventPrepareTransferLeader))
+	}
+}
+
+func transferLeaderCallback(event *fsm.Event) {
+	request := event.Args[0].(*FollowerCallbackRequest)
+	ctx := request.Ctx
+	actionDispatch := request.ActionDispatch
+	c := request.C
+	newLeaderNode := request.NewLeaderNode
+	newLeaderShardID := request.NewLeaderShardID
+
+	// Send event to CeresDB, waiting for response
+	action := dispatch.OpenShardAction{ShardIDs: []uint32{newLeaderShardID}}
+	if err := actionDispatch.OpenShards(ctx, newLeaderNode, action); err != nil {
+		event.Cancel(errors.Wrap(err, EventTransferLeader))
+	}
+
+	if err := updateShardRole(ctx, c, newLeaderShardID, clusterpb.ShardRole_LEADER); err != nil {
+		event.Cancel(errors.Wrap(err, EventPrepareTransferLeader))
+	}
+}
+
+func transferLeaderFailed(event *fsm.Event) {
+	request := event.Args[0].(*FollowerCallbackRequest)
+	ctx := request.Ctx
+	actionDispatch := request.ActionDispatch
+	c := request.C
+	newLeaderNode := request.NewLeaderNode
+	newLeaderShardID := request.NewLeaderShardID
+
+	// Transfer failed, stop transfer and reset state
+	action := dispatch.CloseShardAction{ShardIDs: []uint32{newLeaderShardID}}
+	if err := actionDispatch.CloseShards(ctx, newLeaderNode, action); err != nil {
+		event.Cancel(errors.Wrap(err, EventTransferLeaderFailed))
+	}
+
+	if err := updateShardRole(ctx, c, newLeaderShardID, clusterpb.ShardRole_FOLLOWER); err != nil {
+		event.Cancel(errors.Wrap(err, EventPrepareTransferLeader))
+	}
+}
+
+func updateShardRole(ctx context.Context, c *cluster.Cluster, oldLeaderShardID uint32, role clusterpb.ShardRole) error {
+	shardViews, err := c.GetClusterShardView()
```

Comment:
Maybe we should pass the shard id to the method, and let it return only the necessary shard views.